internal/controllers/reconciliation/patch_test.go

```diff
@@ -127,3 +128,181 @@ func TestPatchDeletion(t *testing.T) {
 	err := downstream.Get(ctx, client.ObjectKeyFromObject(cm), cm)
 	require.True(t, errors.IsNotFound(err))
 }
+
+// TestPatchDeletionBeforeCreation proves that a patch resource can delete the resource it references before the resource is created.
+// Basically, this is the same behavior as Helm hook event with delete policy "before-hook-creation".
+func TestPatchDeletionBeforeCreation(t *testing.T) {
+	scheme := runtime.NewScheme()
+	corev1.SchemeBuilder.AddToScheme(scheme)
+	testv1.SchemeBuilder.AddToScheme(scheme)
+
+	ctx := testutil.NewContext(t)
+	mgr := testutil.NewManager(t)
+	upstream := mgr.GetClient()
+	downstream := mgr.DownstreamClient
+
+	// Test configmap setup
+	cmName := "test-obj"
+	cmNamespace := "default"
+	key := "foo"
+	val := "bar"
+	cm := &corev1.ConfigMap{}
+	cm.Name = cmName
+	cm.Namespace = cmNamespace
+
+	registerControllers(t, mgr)
+	testutil.WithFakeExecutor(t, mgr, func(ctx context.Context, s *apiv1.Synthesizer, input *krmv1.ResourceList) (*krmv1.ResourceList, error) {
+		cm := &unstructured.Unstructured{
+			Object: map[string]any{
+				"apiVersion": "v1",
+				"kind":       "ConfigMap",
+				"metadata": map[string]any{
+					"name":      cmName,
+					"namespace": cmNamespace,
+					"annotations": map[string]string{
+						"eno.azure.io/readiness-group": "2",
+					},
+				},
+				"data": map[string]any{
+					key: val,
+				},
+			},
+		}
+		obj := &unstructured.Unstructured{
+			Object: map[string]any{
+				"apiVersion": "eno.azure.io/v1",
+				"kind":       "Patch",
+				"metadata": map[string]any{
+					"name":      "test-obj",
+					"namespace": "default",
+					"annotations": map[string]string{
+						// This patch should be applied before the configmap is created.
+						"eno.azure.io/readiness-group": "1",
+					},
+				},
+				"patch": map[string]any{
+					"apiVersion": "v1",
+					"kind":       "ConfigMap",
+					"ops": []map[string]any{
+						{"op": "add", "path": "/metadata/deletionTimestamp", "value": "2024-01-22T19:13:15Z"},
+					},
+				},
+			},
+		}
+		return &krmv1.ResourceList{Items: []*unstructured.Unstructured{obj, cm}}, nil
+	})
+
+	// Test subject
+	setupTestSubject(t, mgr)
+	mgr.Start(t)
+	_, comp := writeGenericComposition(t, upstream)
+	testutil.Eventually(t, func() bool {
+		err := upstream.Get(ctx, client.ObjectKeyFromObject(comp), comp)
+		return err == nil && comp.Status.CurrentSynthesis != nil && comp.Status.CurrentSynthesis.Reconciled != nil
+	})
+
+	// Verify the configmap should be created after the deletion patch is applied.
+	err := downstream.Get(ctx, client.ObjectKeyFromObject(cm), cm)
+	require.NoError(t, err)
+	require.NotNil(t, cm)
+	require.Equal(t, val, cm.Data[key])
+}
+
+// TestPatchDeletionBeforeUpgrade proves that a patch resource can delete the resource it references before the resource is upgraded.
+// Basically, this is the same behavior as Helm hook event with delete policy "before-hook-creation".
+func TestPatchDeletionBeforeUpgrade(t *testing.T) {
```

Comment:
Have you tested what happens if either the patch or the resource it deletes are reconciled again after initial reconciliation? i.e. set the reconciliation interval annotation to ~1ms and somehow prove that it doesn't churn?

I suspect that every time the configmap is reconciled it will be recreated, which will trigger reconciliation of the patch since it's in a future readiness group, which will delete it. Not 100% sure though.

Reply:
discuss offline, will test this scenario for the resource reconciliation to see the behavior of deletion patch in next PR